gapis/api/vulkan/framegraph.go

```diff
@@ -166,6 +212,45 @@ type framegraphInfoHelpers struct {
 	rpInfos  []*renderpassInfo
 	rpInfo   *renderpassInfo
 	currRpId uint64
+	// imageLookup is a lookup table to quickly find images that match a memory
+	// access. Scanning the whole state to look for an image matching a memory
+	// access is slow, this lookup table saves seconds of computation. It is
+	// updated when a new renderpass is started under a new top-level parent
+	// command: images can only be created/destroyed between top-level commands,
+	// not during the execution of subcommands.
+	imageLookup  map[memory.PoolID]map[memory.Range][]*ImageObjectʳ
+	parentCmdIdx uint64 // used to know when to update imageLookup
+}
+
+// updateImageLookup updates the lookup table to quickly find an image matching a memory observation.
+func (helpers *framegraphInfoHelpers) updateImageLookup(state *State) {
+	helpers.imageLookup = make(map[memory.PoolID]map[memory.Range][]*ImageObjectʳ)
+	for _, image := range state.Images().All() {
+		for _, aspect := range image.Aspects().All() {
+			for _, layer := range aspect.Layers().All() {
+				for _, level := range layer.Levels().All() {
+					pool := level.Data().Pool()
+					memRange := level.Data().Range()
+					if _, ok := helpers.imageLookup[pool]; !ok {
+						helpers.imageLookup[pool] = make(map[memory.Range][]*ImageObjectʳ)
+					}
+					imgObj := image
+					helpers.imageLookup[pool][memRange] = append(helpers.imageLookup[pool][memRange], &imgObj)
+				}
+			}
+		}
+	}
+}
+
+// lookupImages returns all the images that contain (pool, memRange).
+func (helpers *framegraphInfoHelpers) lookupImages(pool memory.PoolID, memRange memory.Range) []*ImageObjectʳ {
+	images := []*ImageObjectʳ{}
+	for imgRange := range helpers.imageLookup[pool] {
```

Comment:
If we are looping over a map to find if a range is included in a memory range, maybe we should store them as array under a pool instead of a map. Because if we are not using the direct access of the map(and I believe these are hash-tables) then we should not pay the cost for non-consecutive memory access. In a large enough data set, it should be much faster to loop over an array than a map.

Unless, there will be another access to the memory range that will be added in the future that queries the ranges with directly accessing to the map elements.

Reply:
I use a second-level map indexed by ranges in order to avoid storing duplicate ranges. Again, in practice for each poolID we almost always have only a single range anyway.